### ISSUE TITLE
Unusual hard-coded parameters for Adam

### ISSUE THREAD

**Issue by gwenniger** [OpenNMT/OpenNMT-py#382]
It seems to me that the hard-coded default parameters of the Adam optimization 
algorithm in onmt/Optim.py are a bit unusual, as in different from what is being used 
in tensorflow.train.optimizer for example.

The opennmt-py code: 

```
   def set_parameters(self, params):
        self.params = [p for p in params if p.requires_grad]
        if self.method == 'sgd':
            self.optimizer = optim.SGD(self.params, lr=self.lr)
        elif self.method == 'adagrad':
            self.optimizer = optim.Adagrad(self.params, lr=self.lr)
        elif self.method == 'adadelta':
            self.optimizer = optim.Adadelta(self.params, lr=self.lr)
        elif self.method == 'adam':
            self.optimizer = optim.Adam(self.params, lr=self.lr,
                                        betas=self.betas, eps=1e-9)
        else:
            raise RuntimeError("Invalid optim method: " + self.method)
```

    def __init__(self, method, lr, max_grad_norm,
                 lr_decay=1, start_decay_at=None,
                 beta1=0.9, beta2=0.98,
                 opt=None):

Compare to the tensorflow code:

```
__init__(
    learning_rate=0.001,
    beta1=0.9,
    beta2=0.999,
    epsilon=1e-08,
    use_locking=False,
    name='Adam'
)

```
See: https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer

I am trying to replicate the results of 
"Massive Exploration of Neural Machine Translation
Architectures" https://arxiv.org/pdf/1703.03906.pdf
but have a hard time getting to their BLEU scores.

Ironically, a system using the opennmt-py default settings with SGD optimization and feature 
decay gives better results after 12 epochs (22.4 BLEU) than Adam with learning rate 0.001 and no feature decay(after 20 iterations about 21 BLEU) , the latter being a highly popular and successful setup in many recent publications. 
This goes against the established wisdom that Adam supposedly gives better results 
and converges faster than SGD, see for example
"Stronger Baselines for Trustable Results in Neural Machine Translation"
https://arxiv.org/abs/1706.09733

This made me think whether there might be some issue with Adam in opennmt-py, of course 
I'm not sure if this is the case, but it caught my attention that there is this quite large difference in the beta2 parameter used by tensorflow **(0.999)** and the value **0.98** used for Adam in opennmt-py.
In fact 0.999 is also used by Kerras
https://keras.io/optimizers/
this really seems to be the "accepted" value for beta2, not 0.98 as currently in the opennmt-py code.

I haven't yet tried if changing the value to 0.999 gives better results, but I suspect it might, since it seems that that is what everybody else is using.
For example, this default value of tensorflow is also being used in 
https://github.com/google/seq2seq
that gets the high results for English--German translation, and proposes strong parameters.

Also see the original paper proposing Adam
https://arxiv.org/pdf/1412.6980.pdf:

"Good  default  settings  for  the  tested  machine  learning  problems  are
alpha = 0.002 beta1=0.9 and **beta2= 0.999**"

I rest my case. In my opinion,  if the non-standard value of 0.98 is kept it should at least be documented why to use this non-standard value.

**Comment by srush** [comment]
Oh this is great. I would love to standardize this. Could you add these references as a comment to optim, and suggested we changed to use these defaults. Can you send a PR? 

We also were unable to replicate "Stronger Baselines for Trustable Results in Neural Machine Translation", but we would like to. 

The reason we use those beta values [  beta1=0.9, beta2=0.999] is because they were used in the "Attention is All You Need" paper.

**Comment by gwenniger** [comment]
Ok, thanks for your quick reply. I added the changes and made the pull
request.
By the way, did you manage to replicate the English--German results in
https://arxiv.org/abs/1703.03906
As I mentioned, I could not manage to do so yet, I get 22.4 BLEU with the
default
Opennmt settings on their preprocessed data after 12 epochs, and with Adam
using their best parameters as much as possible I get only 21 BLEU after 20
epochs. I am hoping with the alternative version for beta2 the results may
get
better.

Gideon

**Comment by srush** [comment]
Merged.
